sdks/java/transform-service/launcher/src/main/java/org/apache/beam/sdk/transformservice/launcher/TransformServiceLauncher.java

```diff
@@ -0,0 +1,285 @@
+    if (!launchers.containsKey(projectName)) {
+      launchers.put(projectName, new TransformServiceLauncher(projectName, port));
+    }
+    return launchers.get(projectName);
+  }
+
+  private void runDockerComposeCommand(String command) throws IOException {
+    this.runDockerComposeCommand(command, null);
+  }
+
+  private void runDockerComposeCommand(String command, @Nullable File outputOverride)
+      throws IOException {
+    List<String> shellCommand = new ArrayList<>();
+    shellCommand.addAll(dockerComposeStartCommandPrefix);
+    shellCommand.add(command);
+    System.out.println("Executing command: " + String.join(" ", command));
+    ProcessBuilder processBuilder =
+        new ProcessBuilder(shellCommand).redirectError(ProcessBuilder.Redirect.INHERIT);
+
+    if (outputOverride != null) {
+      processBuilder.redirectOutput(outputOverride);
+    } else {
+      processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT);
+    }
+
+    Map<String, String> env = processBuilder.environment();
+    env.putAll(this.environmentVariables);
+
+    processBuilder.start();
+
+    try {
+      this.wait(STATUS_LOGGER_WAIT_TIME);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public synchronized void start() throws IOException, TimeoutException {
+    runDockerComposeCommand("up");
+  }
+
+  public synchronized void shutdown() throws IOException {
+    runDockerComposeCommand("down");
+  }
+
+  public synchronized void status() throws IOException {
+    runDockerComposeCommand("ps");
+  }
+
+  public synchronized void waitTillUp(int timeout) throws IOException, TimeoutException {
+    timeout = timeout <= 0 ? DEFAULT_START_WAIT_TIME : timeout;
+    String statusFileName = getStatus();
+
+    long startTime = System.currentTimeMillis();
+    while (System.currentTimeMillis() - startTime < timeout) {
+      try {
+        // We are just waiting for a local process. No need for exponential backoff.
+        this.wait(1000);
+      } catch (InterruptedException e) {
+        // Ignore and retry.
+      }
+
+      String output = String.join(" ", java.nio.file.Files.readAllLines(Paths.get(statusFileName)));
+      if (!output.isEmpty()) {
+        if (output.contains("transform-service")) {
+          // Transform service was started since we found matching logs.
+          return;
+        }
+      }
+    }
+
+    throw new TimeoutException(
+        "Transform Service did not start in " + timeout / 1000 + " seconds.");
+  }
+
+  private synchronized String getStatus() throws IOException {
+    File outputOverride = File.createTempFile("output_override", null);
+    runDockerComposeCommand("ps", outputOverride);
+
+    return outputOverride.getAbsolutePath();
+  }
+
+  public static void main(String[] args) throws IOException, TimeoutException {
+    String projectName = null;
+    String command = null;
+    int port = -1;
+    String beamVersion = null;
+
+    if (args.length % 2 != 0) {
```

Comment:
this looks a littble bit ad hoc. I imagine it's possible to have single arg in the future. Shall we use some argparse to deal with command line arguments?

Reply:
I don't think there's a single argparse like module for Java. Updated to be bit more generic.

Reply:
We could probably use arg4j which is also used in Beam's flink/spark/samza job servers

Reply:
Updated to use args4j.